yarn-project/epoch-cache/src/epoch_cache.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  EmptyL1RollupConstants,
+  type L1RollupConstants,
+  getEpochNumberAtTimestamp,
+  getSlotAtTimestamp,
+} from '@aztec/circuit-types';
+import {
+  RollupContract,
+  createEthereumChain,
+  getL1ContractsConfigEnvVars,
+  getL1ReaderConfigFromEnv,
+} from '@aztec/ethereum';
+import { EthAddress } from '@aztec/foundation/eth-address';
+
+import { createPublicClient, http } from 'viem';
+
+type EpochAndSlot = {
+  epoch: bigint;
+  slot: bigint;
+  ts: bigint;
+};
+
+export class EpochCache {
+  private validators: Map<EthAddress, boolean>;
+  private currentEpoch: bigint;
+
+  constructor(
+    private rollup: RollupContract,
+    private readonly l1constants: L1RollupConstants = EmptyL1RollupConstants,
+  ) {
+    this.validators = new Map<EthAddress, boolean>();
+
+    this.currentEpoch = getEpochNumberAtTimestamp(BigInt(Math.floor(Date.now() / 1000)), this.l1constants);
+  }
+
+  // TODO: cleanup and merge rollup getters with l1 createAndSync in the archiver
+  static async create(rollupAddress: EthAddress) {
+    const l1ReaderConfig = getL1ReaderConfigFromEnv();
+    const l1constants = getL1ContractsConfigEnvVars();
+
+    const chain = createEthereumChain(l1ReaderConfig.l1RpcUrl, l1ReaderConfig.l1ChainId);
+    const publicClient = createPublicClient({
+      chain: chain.chainInfo,
+      transport: http(chain.rpcUrl),
+      pollingInterval: l1ReaderConfig.viemPollingIntervalMS,
+    });
+
+    const rollup = new RollupContract(publicClient, rollupAddress.toString());
+    const [l1StartBlock, l1GenesisTime] = await Promise.all([
+      rollup.getL1StartBlock(),
+      rollup.getL1GenesisTime(),
+    ] as const);
+
+    const l1RollupConstants: L1RollupConstants = {
+      l1StartBlock,
+      l1GenesisTime,
+      slotDuration: l1constants.aztecSlotDuration,
+      epochDuration: l1constants.aztecEpochDuration,
+      ethereumSlotDuration: l1constants.ethereumSlotDuration,
+    };
+
+    return new EpochCache(rollup, l1RollupConstants);
+  }
+
+  getEpochAndSlotNow(): EpochAndSlot {
+    const now = BigInt(Math.floor(Date.now() / 1000));
```

Comment:
I'm a bit worried that clock drift will cause issues in this setup, but do not want to poll the smart contract for everything, what do we think?

Reply:
Agreed. We could have it reset at a preconfigured interval? e.g. once per epoch sync with the rollup.

Reply:
Filed https://github.com/AztecProtocol/aztec-packages/issues/10435